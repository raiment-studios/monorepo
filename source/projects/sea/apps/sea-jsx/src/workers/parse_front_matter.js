import yaml from 'yaml';

/**
 * Scan the entry-point file for front-matter data that will be used in
 * configuration.
 *
 * @param {*} source
 *
 * TODO: make this more efficient. It is currently written for simplicity
 * with no thought to efficiency.
 */
export function parseFrontMatter(source) {
    const lines = source.split('\n');
    const start = lines.findIndex((line) => line.trim().startsWith('/*!@sea:header'));

    // There's no front-matter defined
    if (start === -1) {
        return null;
    }

    const end = lines.slice(start + 1).findIndex((line) => line.trim() === '*/');
    if (end === -1) {
        console.error('Error: found front-matter begin token without end token');
        process.exit(1);
    }

    const content = lines.slice(start + 1, start + end + 1).join('\n');
    const obj = yaml.parse(content);
    return obj;
}
