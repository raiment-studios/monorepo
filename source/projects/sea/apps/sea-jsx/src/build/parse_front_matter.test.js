import { describe, expect, test } from '@jest/globals';
import { parseFrontMatter } from './parse_front_matter';

describe('parseFrontMatter', () => {
    test('lack of front matter returns null', () => {
        const obj = parseFrontMatter('');
        expect(obj).toBe(null);
    });

    const testCases = {
        'empty string': '',
        'no comments': `
function hello() {
    return 'world';
}
        `,
        'no comment header': `
/*!
    hello: world
*/
function hello() {
    return 'world';
}
        `,
        indented: `
/*!@sea:header
    hello: world
*/
function hello() {
    return 'world';
}
        `,
        'arbitrary YAML is parsed': `
/*!@sea:header
hello: world
list:
    - 1
    - 2
    - 3
    - nested1: value1
      nested2: value2
*/
        `,
    };

    for (let [name, text] of Object.entries(testCases)) {
        test(name, () => {
            const result = parseFrontMatter(text);
            expect(result).toMatchSnapshot();
        });
    }
});
