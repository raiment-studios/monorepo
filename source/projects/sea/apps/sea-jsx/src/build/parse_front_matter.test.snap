// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`parseFrontMatter arbitrary YAML is parsed 1`] = `
Object {
  "hello": "world",
  "list": Array [
    1,
    2,
    3,
    Object {
      "nested1": "value1",
      "nested2": "value2",
    },
  ],
}
`;

exports[`parseFrontMatter empty string 1`] = `null`;

exports[`parseFrontMatter indented 1`] = `
Object {
  "hello": "world",
}
`;

exports[`parseFrontMatter no comment header 1`] = `null`;

exports[`parseFrontMatter no comments 1`] = `null`;
